### PR TITLE
add a tde demo

### PR DESCRIPTION
**Pull request by Sasasu**

### CODE REVIEW

File: src/bin/initdb/initdb.c [hunk 1]
```diff
@@ -2655,6 +2655,7 @@ usage(const char *progname)
 	printf(_("\nLess commonly used options:\n"));
 	printf(_("  -d, --debug               generate lots of debugging output\n"));
 	printf(_("  -k, --data-checksums      use data page checksums\n"));
+	printf(_("  --preload_libraries       load libraries, same as shared_preload_libraries in postgresql.conf\n"));
```
Comment (higuoxing) [inline src/bin/initdb/initdb.c]: Let's keep the argument's name consistent with other arguments. `--preload-libraries`.

File: gpcontrib/data_invert/Makefile [hunk 2]
```diff
@@ -0,0 +1,15 @@
+PG_CONFIG = pg_config
```
Comment (higuoxing) [inline gpcontrib/data_invert/Makefile]: Seems that this variable is not used.

File: gpMgmt/bin/gpinitsystem [hunk 3]
```diff
@@ -1373,6 +1377,7 @@ CREATE_SEGMENT () {
 			# SEGMENT_INFO: ~ separated value, ex: host1~port1~data_dir~dbid1~contentid
 			# PAIR_SEGMENT_INFO: corresponding pair, ex: host2~port2~data_dir~dbid2~contentid
 			# INST_COUNT: Number of parallel script count
+			export PRELOAD_LIBRARIES
```
Comment (higuoxing) [inline gpMgmt/bin/gpinitsystem]: Can we pass it as an argument to make the operation idempotent?

File: src/backend/utils/init/miscinit.c [hunk 4]
```diff
@@ -1688,6 +1688,17 @@ process_shared_preload_libraries(void)
...
 	process_shared_preload_libraries_in_progress = true;
 
+	if ((shared_preload_libraries_string == NULL || strlen(shared_preload_libraries_string) == 0)
+		&& (IsBootstrapProcessingMode() || IsInitProcessingMode()))
+	{
+		/*
+		 * if we are in init or bootstrap mode. postgresql.conf does not exists yet.
+		 * try to load libraries from os environment
+		 */
+		char * const env_preload = getenv("GP_PRELOAD_LIBRARIES");
```
Comment (higuoxing) [inline src/backend/utils/init/miscinit.c:1698]: What's the difference between `GP_PRELOAD_LIBRARIES` and `PRELOAD_LIBRARIES`?

Comment (Sasasu) [reply src/backend/utils/init/miscinit.c:1698]: `GP_PRELOAD_LIBRARIES` is an environment variable, `PRELOAD_LIBRARIES` is a shell variable (but pass in to create_segment.sh using environment)

File: gpMgmt/bin/gpinitsystem (hunk 3 above)
Comment (Sasasu) [reply gpMgmt/bin/gpinitsystem]: There is an argument for gpinitsystem. But leak an argument for `make create-demo-cluster`

Comment (higuoxing) [reply gpMgmt/bin/gpinitsystem]: I'm afraid exporting additional environment variables is making things complex. What do you think of adding an additional argument `PRELOAD_LIBRARIES` to the make command. E.g., `make PRELOAD_LIBRARIES=data_invert create-demo-cluster`

Comment (Sasasu) [reply gpMgmt/bin/gpinitsystem]: There are 3 data flows:

1. initdb to postgresql: using GP_PRELOAD_LIBRARIES env
2. gpinitsystem to gpcreateseg.sh: using PRELOAD_LIBRARIES env (renamed to GP_PRELOAD_LIBRARIES now)
3. from `make create-demo-cluster` to gpinitsystem: hard-coded for now.

I will add some like `GP_PRELOAD_LIBRARIES=data_invert make create-demo-cluster` after I fix all error in `make installcheck`

File: gpAux/gpdemo/demo_cluster.sh [hunk 5]
```diff
@@ -395,6 +395,8 @@ fi
 
 # Add fsync=off for all gpdemo deployments
 grep -q 'fsync=off' ${CLUSTER_CONFIG_POSTGRES_ADDONS} && echo "fsync=off already exists in ${CLUSTER_CONFIG_POSTGRES_ADDONS}." || echo "fsync=off" >> ${CLUSTER_CONFIG_POSTGRES_ADDONS}
+echo 'shared_preload_libraries=data_invert' >> ${CLUSTER_CONFIG_POSTGRES_ADDONS}
```
Comment (beeender) [inline gpAux/gpdemo/demo_cluster.sh]: let's use `file_hook_test` instead?

**Comment by Sasasu** [comment]
CI
https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-sa_tde-rocky8

File: src/backend/storage/smgr/smgr.c [hunk 6]
```diff
@@ -581,7 +581,7 @@ smgrextend(SMgrRelation reln, ForkNumber forknum, BlockNumber blocknum,
 		   char *buffer, bool skipFsync)
 {
 	if (file_extend_hook)
-		(*file_extend_hook)(reln, forknum, blocknum, buffer);
+		buffer = (*file_extend_hook)(reln, forknum, blocknum, buffer);
```
Comment (beeender) [inline src/backend/storage/smgr/smgr.c:584]: When will the returned `buffer` be freed?

Comment (Sasasu) [reply src/backend/storage/smgr/smgr.c:584]: This is a thread local, or long live in process memory.
```
// process local buffer
static char file_write_hook_buffer[BLCKSZ]                       = {'\0'};
static char ao_file_write_hook_buffer[MAX_APPENDONLY_BLOCK_SIZE] = {'\0'};
```

no need to free

Comment (beeender) [reply src/backend/storage/smgr/smgr.c:584]: This needs to be documented where `file_extend_hook` is declared. Do we want to have a `bool* shouldfree` in the functions' signature? like:

```
typedef char* (*file_extend_hook_type)(SMgrRelation reln, ForkNumber forknum, BlockNumber blocknum, char *buffer, bool* shouldfree);
```

Comment (beeender) [reply src/backend/storage/smgr/smgr.c:584]: hmm... `file_write_hook` has the same situation. I feel a `bool* sholdfree` is not necessary, just add comment to say:

```
The returned pointer won't be freed after calling the hook. Hook implementation should ensure the returned buffer is valid and won't be leaked.
```

File: src/bin/initdb/initdb.c [hunk 7]
```diff
@@ -3490,6 +3492,9 @@ main(int argc, char *argv[])
 			case 1003:
 				n_buffers = parse_long(optarg, true, "shared_buffers");
 				break;
+			case 1004:
+				setenv("GP_PRELOAD_LIBRARIES", optarg, true);
```
Comment (higuoxing) [inline src/bin/initdb/initdb.c:3496]: Can we pass the flag via command line options? Mutating the global environment seems not a good idea. Is there any specific reason for that.

**Pull request closed without merging**